File: matlab/lib/+IceInternal/ValueFactoryManagerI.m
%{
**********************************************************************

Copyright (c) 2003-2018 ZeroC, Inc. All rights reserved.

This copy of Ice is licensed to you under the terms described in the
ICE_LICENSE file included in this distribution.

**********************************************************************
%}

classdef ValueFactoryManagerI < Ice.ValueFactoryManager
    methods
        function obj = ValueFactoryManagerI()
            obj.factories = containers.Map('KeyType', 'char', 'ValueType', 'any');
            obj.hasFactories = false;
            obj.defaultFactory = [];
        end
        function add(obj, factory, id)
            if isempty(id)
                if ~isempty(obj.defaultFactory)
                    throw(Ice.AlreadyRegisteredException('', '', 'value factory', id));
                end
                obj.defaultFactory = factory;
            else
                if obj.factories.isKey(id)
                    throw(Ice.AlreadyRegisteredException('', '', 'value factory', id));
                elseif ~isa(factory, 'function_handle')
                    throw(MException('Ice:ArgumentException', 'factory argument is not a function handle'));
                else
                    obj.factories(id) = factory;
                    obj.hasFactories = true;
                end
            end
        end
        function r = find(obj, id)
            if isempty(id)
                r = obj.defaultFactory;
            elseif obj.hasFactories && obj.factories.isKey(id)
                r = obj.factories(id);
            else
                r = [];
            end
        end
    end
    properties(Access=private)
        factories
        hasFactories
        defaultFactory
    end
end

File: java/src/Ice/src/main/java/com/zeroc/Ice/ValueFactoryManagerI.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

package com.zeroc.Ice;

final class ValueFactoryManagerI implements ValueFactoryManager {
    public synchronized void add(ValueFactory factory, String id) {
        if (_factoryMap.containsKey(id)) {
            throw new AlreadyRegisteredException("value factory", id);
        }
        _factoryMap.put(id, factory);
    }

    public synchronized ValueFactory find(String id) {
        return _factoryMap.get(id);
    }

    private final java.util.Map<String, ValueFactory> _factoryMap = new java.util.HashMap<>();
}
